### PR TITLE
Documentation for DataStore Hub Events

### PR DESCRIPTION
**Pull request by ruiguoamz**
*Issue #, if available:* #2379 

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: docs/lib/datastore/fragments/ios/datastore-events.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
+## Usage
+For instance, to see if the network status is active, you could set up the following listener or Subscriber (using Combine):
```
Comment (drochetti) [inline docs/lib/datastore/fragments/ios/datastore-events.md:2]: I think a more introductory paragraph is missing. Introducing what hub events are and what use cases they cover.

Also, this content can/should be moved to the `docs/lib/datastore/datastore-events.md` file since it's common to all platforms.

File: docs/lib/datastore/fragments/ios/datastore-events.md [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+}
+```
+
+</amplify-block>
+
+</amplify-block-switcher>
+
+<amplify-callout>
+
+For `networkStatus` event, an initial value in which `active` is always set to `false` is received, but shortly after, you will receive a updated value which represents the real status of network connectivity.
```
Comment (drochetti) [inline docs/lib/datastore/fragments/ios/datastore-events.md:42]: hhmm let's discuss this. This would represent a platform implementation difference and can be quite confusing.

Is there a mitigation we can put in our implementation?

File: docs/lib/datastore/fragments/native_common/datastore-events.md [hunk 3]
```diff
@@ -1,14 +1,102 @@
 Amplify uses `Hub` for different categories to communicate with one another when specific events occur. DataStore emits `Hub` messages for the following events:
 
+## networkStatus
+
+Dispatched when DataStore starts and every time network status changes
+
+HubPayload `NetworkStatusEvent` contains a boolean value `active` to notify network status
+```
+{ active: true }
+```
```
Comment (drochetti) [inline docs/lib/datastore/fragments/native_common/datastore-events.md:10]: I think putting JSON blocks doesn't make much sense from an iOS and Android perspective. I think the list structure you have below in the `modelSynced` could be used for all events.

File: docs/lib/datastore/fragments/native_common/datastore-events.md [hunk 4]
```diff
@@ -1,14 +1,102 @@
...
+
+## subscriptionsEstablished
+
+Dispatched when DataStore has finished establishing its subscriptions to all sync-able models
```
Comment (drochetti) [inline docs/lib/datastore/fragments/native_common/datastore-events.md:14]: what is "sync-able" from a customer perspective? I think we can omit it.

File: docs/lib/datastore/fragments/native_common/datastore-events.md [hunk 5]
```diff
@@ -1,14 +1,102 @@
...
+
+Dispatched once for each model after the model instances have been synced from the cloud.
+
+HubPayload `modelSyncedEvent` contains:
+- `modelName` (String): the name of the model that was synced
+- `isFullSync` (Bool): `true` if the model was synced with a "full" query to retrieve all models
+- `isDeltaSync` (Bool): `true` if the model was synced with a "delta" query to retrieve only changes since the last sync
+- `create` (Int): the number of new model instances added to the local store
+- `update` (Int): the number of existing model instances updated in the local store
+- `delete` (Int): the number of model instances deleted from the local stor
```
Comment (drochetti) [inline docs/lib/datastore/fragments/native_common/datastore-events.md:37]: I like this format, maybe we can use it as the standard for all event?

**Review by drochetti — COMMENTED** [review]
thanks for doing this, I think this is a much improved version compared with what we have today. Let's iterate a bit more to establish a good bar for hub events documentation.

**Comment by drochetti** [comment]
This PR is related to #2379

File: docs/lib/datastore/fragments/native_common/datastore-events.md (hunk 3 above)
Comment (ruiguoamz) [reply docs/lib/datastore/fragments/native_common/datastore-events.md]: Updated

File: docs/lib/datastore/fragments/native_common/datastore-events.md (hunk 4 above)
Comment (ruiguoamz) [reply docs/lib/datastore/fragments/native_common/datastore-events.md]: removed

File: docs/lib/datastore/fragments/native_common/datastore-events.md [hunk 6]
```diff
@@ -1,14 +1,87 @@
...
+* There are outgoing changes not yet synchronized
+
+There are following nine DataStore events:
+
+## networkStatus
+
+Dispatched when DataStore starts and every time network status changes
+
+HubPayload `NetworkStatusEvent` contains:
+- `active` (Bool): a boolean value to notify network status
```
Comment (jamesonwilliams) [inline docs/lib/datastore/fragments/native_common/datastore-events.md:15]:
```suggestion
- `active` (Bool): true if the DataStore is on a network that can connect the Cloud; false, otherwise
```

**Review by jamesonwilliams — COMMENTED** [review]
Good start!

I have included some attempts to simplify the wording, though English remains hard for me after speaking it for 34 years. 😭 

Are the names of all of these events the same as on iOS? Right now, there is a `native-common` path referenced in this PR. That means that the content must be identically accurate for Android, too. @rjuliano, would you mind to double the check the event names/descriptions for consistency with Android?

File: docs/lib/datastore/fragments/native_common/datastore-events.md [hunk 7]
```diff
@@ -1,14 +1,87 @@
...
+
+Dispatched when a mutation from outgoing mutation queue `outbox` is sent to backend and updated locally.
+
+HubPayload `outboxMutationEvent` contains: the name and instance of the model
+- `modelName` (String): the model name of the mutation that was enqueued,
+- `element`: 
+    - `model` (Model): the instance of your model
+    - `_version` (Int): version of the model instance
+    - `_lastChangedAt` (Int): last change time of model instance (epoch time)
+    - `_deleted` (Bool): whether the model instance is deleted or not
```
[the remark this answers is not in the document]
Comment (ruiguoamz) [reply docs/lib/datastore/fragments/native_common/datastore-events.md:74]: For `_deleted` is actually whether the data is deleted in backend, I changed it to 
```
    - `_deleted` (Bool): true if the model instance has been deleted in Cloud
```

File: docs/lib/datastore/fragments/native_common/datastore-events.md [hunk 8]
```diff
@@ -1,14 +1,87 @@
...
+## outboxMutationProcessed
+
+Dispatched when a mutation from outgoing mutation queue `outbox` is sent to backend and updated locally.
+
+HubPayload `outboxMutationEvent` contains: the name and instance of the model
+- `modelName` (String): the model name of the mutation that was enqueued,
+- `element`: 
+    - `model` (Model): the instance of your model
+    - `_version` (Int): version of the model instance
+    - `_lastChangedAt` (Int): last change time of model instance (epoch time)
```
[the remark this answers is not in the document]
Comment (ruiguoamz) [reply docs/lib/datastore/fragments/native_common/datastore-events.md:73]: I changed the `epoch time` to `unix time` which is always represented by seconds

File: docs/lib/datastore/fragments/native_common/datastore-events.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (jamesonwilliams) [reply docs/lib/datastore/fragments/native_common/datastore-events.md]: Whoops, good call. Thanks!

File: docs/lib/datastore/fragments/js/datastore-events.md [hunk 9]
```diff
@@ -1,5 +1,3 @@
-For instance, to listen to see if the network status is active, you could set up the following listener:
```
Comment (sammartinez) [inline docs/lib/datastore/fragments/js/datastore-events.md:1]: What is the need for removing this? Are we placing a fragment here?

Comment (ruiguoamz) [reply docs/lib/datastore/fragments/js/datastore-events.md:1]: It was a change according to Daniel's suggestion. see [this comment](https://github.com/aws-amplify/docs/pull/2470#discussion_r497881887)

Comment (sammartinez) [reply docs/lib/datastore/fragments/js/datastore-events.md:1]: @drochetti Can you weigh in on this? I don't want to remove the line from JS if there is no replacement for it. I am seeing that the iOS and Android use cases have the extend paragraph but if we aren't going to add it for JS I dont want this removed.

Comment (drochetti) [reply docs/lib/datastore/fragments/js/datastore-events.md:1]: My suggestion in that comment was to add an introductory paragraph and then share the content across _all_ platforms.

I see the changes are only shared in a `native_common` fragment. Any reason not to allow JS to use the same content @ruiguoamz?

Comment (ruiguoamz) [reply docs/lib/datastore/fragments/js/datastore-events.md:1]: I don't think there is reason not to allow JS to use the same content. The description should be the same for 3 platforms, only the code snippets are different

File: docs/lib/datastore/fragments/native_common/datastore-events.md [hunk 10]
```diff
@@ -1,14 +1,84 @@
...
-outboxMutationProcessed
-outboxStatus
-networkStatus
-ready
-```
\ No newline at end of file
+DataStore periodically publishes state notifications onto Amplify's Hub. You can subscribe to the Hub to gain insight into the internal state of the DataStore. Events are published when:
+* Your device loses or regains network connectivity;
+* Data is synchronized with the Cloud;
+* There are new, pending changes that are have not yet been synchronized.
```
Comment (jamesonwilliams) [inline docs/lib/datastore/fragments/native_common/datastore-events.md:4]:
```suggestion
* There are new, pending changes that have not yet been synchronized.
```

File: docs/lib/datastore/fragments/native_common/datastore-events.md [hunk 11]
```diff
@@ -1,14 +1,84 @@
...
+Dispatched when DataStore starts and every time network status changes
+
+HubPayload `NetworkStatusEvent` contains:
+- `active` (Bool): true if the DataStore is on a network that can connect the Cloud; false, otherwise
+
+## subscriptionsEstablished
+
+Dispatched when DataStore has finished establishing its subscriptions to all models
+
+HubPayload: `N/A`
```
Comment (jamesonwilliams) [inline docs/lib/datastore/fragments/native_common/datastore-events.md:18]:
```suggestion
HubPayload: N/A
```

File: docs/lib/datastore/fragments/native_common/datastore-events.md [hunk 12]
```diff
@@ -1,14 +1,84 @@
...
+
+Dispatched when all models have been synced from the cloud
+
+HubPayload: `N/A`
+
+## ready
+
+Dispatched when DataStore as a whole is ready, at this point all data is available
+
+HubPayload: `N/A`
```
Comment (jamesonwilliams) [inline docs/lib/datastore/fragments/native_common/datastore-events.md:49]:
```suggestion
HubPayload: N/A
```

**Review by jamesonwilliams — APPROVED** [review]

**Review by drochetti — APPROVED** [review]
🚀

**Review by manueliglesias — APPROVED** [review]
LGTM 👍
